src/ComplexType.php

```diff
@@ -101,93 +120,228 @@ protected function generateClass()
+     * @param $typeHint
+     * @param $defaultToNull
+     */
+    private static function addMethodParameter(MethodGenerator $method, $name, $type, $typeHint, $defaultToNull)
+    {
+        $parameter = new ParameterGenerator();
+        $parameter->setName($name);
+
+        $method->getDocBlock()->setTag(new ParamTag($name, $type));
+
+        if (!empty($typeHint)) {
+            $parameter->setType($typeHint);
+        }
+        if ($defaultToNull) {
+            $parameter->setDefaultValue(
+                new ValueGenerator(null, ValueGenerator::TYPE_NULL)
             );
-            $accessors[] = $setter;
         }
+        $method->setParameter($parameter);
+    }
+
+    /**
+     * Add inherited parameter to constructor
+     *
+     * @param MethodGenerator $constructor
+     * @param $name
+     * @param $type
+     * @param $typeHint
+     * @param $constructorParamsDefaultToNull
+     * @return ParameterGenerator
+     */
+    private static function addParentConstructorParameter(
+        MethodGenerator $constructor,
+        $name, $type, $typeHint, $constructorParamsDefaultToNull
+    )
+    {
+        self::addMethodParameter($constructor, $name, $type, $typeHint, $constructorParamsDefaultToNull);
+
+        $parentConstructorParameter = new ParameterGenerator();
+        $parentConstructorParameter->setName($name);
+        if (!empty($typeHint)) {
+            $parentConstructorParameter->setType($typeHint);
+        }
+
+        return $parentConstructorParameter;
+    }
+
+    /**
+     * Add parameter to constructor
+     *
+     * @param MethodGenerator $constructor
+     * @param $name
+     * @param $type
+     * @param $typeHint
+     * @param $nullable
+     * @return string Source code to add to constructor
+     */
+    private static function addConstructorParameter(
+        MethodGenerator $constructor,
+        $name, $type, $typeHint, $nullable, $constructorParamsDefaultToNull)
+    {
+        self::addMethodParameter($constructor, $name, $type, $typeHint, $constructorParamsDefaultToNull);
 
-        $constructor = new PhpFunction(
-            'public',
-            '__construct',
-            $this->buildParametersString(
-                $constructorParameters,
-                true,
-                $this->config->get('constructorParamsDefaultToNull')
-            ),
-            $constructorSource,
-            $constructorComment
+        $constructorSource = self::generateSetterSource(
+            $name, $type, $nullable || $constructorParamsDefaultToNull
         );
-        $this->class->addFunction($constructor);
 
-        foreach ($accessors as $accessor) {
-            $this->class->addFunction($accessor);
+        return $constructorSource;
+    }
+
+    /**
+     * Add property to generated class
+     *
+     * @param string $name
+     * @param string $type
+     */
+    private function addProperty($name, $type)
+    {
+        $property = new PropertyGenerator();
+        $property
```

Comment:
Could we leave out this line and call methods directly on the new `PropertyGenerator`?